src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs

```diff
@@ -49,17 +51,24 @@ internal class RoslynVisualStudioWorkspace : VisualStudioWorkspaceImpl
 
         internal override IInvisibleEditor OpenInvisibleEditor(DocumentId documentId)
         {
-            var globalUndoService = this.Services.GetService<IGlobalUndoService>();
+            var globalUndoService = this.Services.GetRequiredService<IGlobalUndoService>();
             var needsUndoDisabled = false;
 
+            var textDocument = this.CurrentSolution.GetTextDocument(documentId);
+
+            if (textDocument == null)
+            {
+                throw new ArgumentException(string.Format(WorkspacesResources._0_is_not_part_of_the_workspace, documentId), nameof(documentId));
```

Comment:
>hrow new ArgumentException( [](start = 17, length = 27)

Why not use Contract.ThrowIfNull()? Is this called from a public API?

Reply:
It's public-ish -- I suspect it's being called via IVT. And we already had a string so this would make debugging a bit easier since the dump would have a more useful exception message.

Reply:
I'd leave ArgumentException for public API validation only. If you think the message is useful, I'd use InvalidOperationException.

Reply:
I guess I don't see a reason to change exception for internal vs. public validation....?

Reply:
The fact that you're throwing `ArgumentException` indicates that this method is supposed to do full argument validation as if it was public API. This also means the exceptions thrown are part of the public contract of the method and should not be changed.

If you just do `Contract.ThrowIfNull` then that indicates this method is doing internal contract validation that doesn't have such a high bar as public API.

Reply:
Fixed. I guess this makes sense since if the exception did leak since it's an inner exception people can't (sanely) catch the exception type.